contracts/vault/VaultInventoryReporter.sol

```diff
@@ -0,0 +1,606 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.11;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "@openzeppelin/contracts/token/ERC1155/IERC1155.sol";
+import "@openzeppelin/contracts/utils/cryptography/draft-EIP712.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableMap.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./VaultOwnershipChecker.sol";
+import "./OwnableERC721.sol";
+import "../interfaces/IVaultInventoryReporter.sol";
+import "../external/interfaces/IPunks.sol";
+
+/**
+ * @title VaultInventoryReporter
+ * @author Non-Fungible Technologies, Inc.
+ *
+ * The VaultInventoryReporter contract is a global tracker of reported
+ * inventory in all Arcade Asset Vaults. This reporting should _always_
+ * be accurate, but will _not_ be comprehensive - that is, many vaults
+ * will end up having unreported inventory. This contract should
+ * be used specifically to verify _whether_ certain items are in the
+ * reported inventory, and not to get a sense of truth as to _all_
+ * the items in a particular vault.
```

Comment:
Possibly expand on this a bit describing how if an item is removed directly from a vault and not through the `remove` functionality, it results in a 'stale' reading when enumerating, still showing the item in inventory even though has been removed externally through a direct interaction with the vault.

Reply:
Well, the `remove` functionality doesn't actually remove assets from the vault either, it just updates reporting. So yeah withdrawing and updating will always be separate steps, especially because the `VaultDepositRouter` isn't concerned with withdrawing.

In practice, by the time a user calls `withdrawEnabled`, the reporting shouldn't matter anyway, since that vault can no longer be used in a loan or transferred at all. I think the more likely use case of `remove` / `clear` would be to remove something from inventory that was mistakenly added and _never_ held by the vault.